### PR TITLE
Day to Night Mode

### PR DESCRIPTION
**Pull request by sracca**
- Added DS store to .gitignore file
- Add Solarcycle ActiveRecord to keep track of sunrise and sunset at all time.
- Changed javascript to reflect this data and utilize this instead of daily weather data
- Update view to reflect what is seen on zepplin + change circle icon based on current day's moonPhase

Screen Shot 2019-06-04 at 1 06 35 PM

### CODE REVIEW

File: app/assets/javascripts/solarcycles.coffee [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Place all the behaviors and hooks related to the matching controller here.
+# All this logic will automatically be available in application.js.
+# You can use CoffeeScript in this file: http://coffeescript.org/
```
Comment (geraldalewis) [inline app/assets/javascripts/solarcycles.coffee]: `can` use CoffeeScript ... 😭 

image

(...we should prob remove these auto-generated files?)

File: app/graphql/types/solarcycle_type.rb [hunk 2]
```diff
@@ -1,8 +1,8 @@
 
 
 class Types::SolarcycleType < Types::BaseObject
-    graphql_name "SolarCycle"
-  
-    field "type", String
-    field "time", Types::UnixDateTimeType
-  end
\ No newline at end of file
+  graphql_name "SolarCycle"
```
Comment (geraldalewis) [inline app/graphql/types/solarcycle_type.rb]: Could be wrong here, but this seems to belong to another commit? Or maybe the other commit should be combined here?

File: app/lib/clients/darksky_client.rb [hunk 3]
```diff
@@ -1,9 +1,29 @@
 class Clients::DarkskyClient
-  def self.forecast(latitude, longitude)
-    key = "darksky-response-#{latitude}-#{longitude}"
-
+  def self.forecast(location)
+    key = "darksky-response-#{location.latitude}-#{location.longitude}"
     Rails.cache.fetch(key, expires_in: 5.minutes) do
-      ForecastIO.forecast(latitude, longitude)
+      forecast = ForecastIO.forecast(location.latitude, location.longitude)
+      debugger
```
Comment (geraldalewis) [inline app/lib/clients/darksky_client.rb]: ^

File: app/workers/weather_poller_worker.rb [hunk 4]
```diff
@@ -9,32 +9,36 @@ def perform
...
-  def poll(latitude, longitude)
-    logger.info "Polling #{latitude} #{longitude}"
+  def poll(location)
+    logger.info "Polling #{location.latitude} #{location.longitude}"
+    location_record = Location.where(latitude: location.latitude, longitude: location.longitude).first
+    get_forecast(location_record, location)
+  end
 
-    data = Clients::DarkskyClient.forecast(latitude, longitude)
+  def get_forecast(record, location)
```
Comment (geraldalewis) [inline app/workers/weather_poller_worker.rb]: Nice 👍

File: app/lib/clients/darksky_client.rb [hunk 5]
```diff
@@ -3,7 +3,6 @@ def self.forecast(location)
     key = "darksky-response-#{location.latitude}-#{location.longitude}"
     Rails.cache.fetch(key, expires_in: 5.minutes) do
       forecast = ForecastIO.forecast(location.latitude, location.longitude)
-      debugger
```
Comment (geraldalewis) [inline app/lib/clients/darksky_client.rb]: (folds into prev commit)

File: spec/workers/weather_poller_worker_spec.rb [hunk 6]
```diff
@@ -1,39 +1,58 @@
 require 'rails_helper'
+
+TESTDATA =
+  {
+    "daily" => {
+      "data" => [
+        {
+          "sunsetTime" => 1_559_573_189,
+          "sunriseTime" => 1_559_573_189
```
Comment (geraldalewis) [inline spec/workers/weather_poller_worker_spec.rb]: Even tho it's test data, can we make `sunsetTime > sunriseTime` (or vice-versa if appropriate)?

File: app/javascript/components/widgets/weather/sunrise-sunset.jsx [hunk 7]
```diff
@@ -53,62 +53,60 @@ const SunsetLabel = styled.div`
...
+const SunriseSunset = ({ location }) => {
+  const { timezone, solarcycles } = location;
+  const currDate = new Date();
+
+  solarcycles.forEach(element => {
+    console.log(element);
+  });
+  const startTime = solarcycles
+    .filter(cycle => new Date(cycle.time).getTime() - currDate.getTime() < 0)
+    .slice(-1)[0];
```
Comment (geraldalewis) [inline app/javascript/components/widgets/weather/sunrise-sunset.jsx]: Pre-empting @cbarber with `takeLast`?

File: app/javascript/components/widgets/weather/semi-circle.jsx [hunk 8]
```diff
@@ -11,7 +11,7 @@ import { colors } from '../../../lib/theme';
 export class SemiCircle extends React.Component {
   static propTypes = {
     totalTime: PropTypes.number.isRequired,
-    sunset: PropTypes.instanceOf(Date).isRequired,
+    endTime: PropTypes.instanceOf(Date).isRequired,
```
Comment (geraldalewis) [inline app/javascript/components/widgets/weather/semi-circle.jsx]: Curious about the name change -- what were your thoughts about it?

File: app/javascript/components/widgets/weather/moon-phase.jsx [hunk 9]
```diff
@@ -0,0 +1,132 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { colors } from '../../../lib/theme';
+
+const iconRender = (posX, posY, radius, phase) => {
+  switch (true) {
```
Comment (geraldalewis) [inline app/javascript/components/widgets/weather/moon-phase.jsx]: I think we want

```javascript
switch (phase) {
  case 'newMoon':
    ....
}
```

File: app/javascript/components/widgets/weather/moon-phase.jsx [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+            fill={colors.black}
+            stroke={colors.black}
+            mask={`url(#${phase})`}
+          />
+        </g>
+      );
+  }
+};
+
+const MoonPhase = ({ posX, posY, radius, phase }) => (
```
Comment (geraldalewis) [inline app/javascript/components/widgets/weather/moon-phase.jsx]: 🔥 🔥 🔥 👍

File: app/javascript/components/widgets/weather/semi-circle.jsx [hunk 11]
```diff
@@ -7,6 +7,28 @@ import {
   timeAndDateForTimezone,
 } from '../../../lib/datetime';
 import { colors } from '../../../lib/theme';
+import MoonPhase from './moon-phase';
+
+const getPhaseType = phase => {
+  switch (true) {
+    case phase === 0:
```
Comment (geraldalewis) [inline app/javascript/components/widgets/weather/semi-circle.jsx]: `Java` has the restriction that the `switch` condition must be an `int` (and now `int | String`) but JS will take whatever kind of value you throw at it (tho it won't do deep equality checks for `Object` values).

File: helios2.code-workspace [hunk 12]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (geraldalewis) [inline helios2.code-workspace]: New commit (or _prob_ ditch this)?

File: app/javascript/components/widgets/weather/semi-circle.jsx (hunk 8 above)
Comment (sracca) [reply app/javascript/components/widgets/weather/semi-circle.jsx]: I wasn't a hundred percent happy changing it to start and end (thought it was too vague), but sunrise and sunset weren't accurate/confusing and I honestly couldn't think of what would be a better name at the time. Suggestions are welcome!

File: helios2.code-workspace (hunk 12 above)
Comment (sracca) [reply helios2.code-workspace]: Whoops, yeah, I thought I had gotten rid of that

File: app/javascript/components/widgets/weather/moon-phase.jsx (hunk 9 above)
Comment (sracca) [reply app/javascript/components/widgets/weather/moon-phase.jsx]: I use includes() to reduce repetition for the Gibbous and Quarter shapes, how do you recommend I convert that to this style?

File: db/schema.rb [hunk 13]
```diff
@@ -32,6 +24,14 @@
     t.index ["location_id"], name: "index_announcements_on_location_id"
   end
 
+  create_table "daily_event_summaries", force: :cascade do |t|
```
[the remark this answers is not in the document]
Comment (sracca) [reply db/schema.rb]: Schema.rb is auto-generated on command. It may have switched order when re-creating it.

File: app/javascript/components/widgets/weather/sunrise-sunset.jsx (hunk 7 above)
Comment (cbarber) [reply app/javascript/components/widgets/weather/sunrise-sunset.jsx]: Close! I think you can probably name things a little better:
```
const [beforeNow, afterNow] = slitWhen(cycle => ....).;
const startTime = takeLast(1, afterNow);
const stopTime = take(1,afterNow), 
```

This includes a pet peeve of mine. Always start/stop or begin/end. Never start/end or begin/stop :)

File: app/javascript/components/widgets/weather/sunrise-sunset.jsx [hunk 14]
```diff
@@ -53,62 +53,60 @@ const SunsetLabel = styled.div`
...
 
 SunriseSunset.propTypes = {
-  weather: PropTypes.shape({
-    daily: PropTypes.shape({
-      data: PropTypes.array.isRequired,
-    }).isRequired,
-  }).isRequired,
   location: PropTypes.shape({
     timezone: PropTypes.string.isRequired,
+    solarcycles: PropTypes.array.isRequired,
```
Comment (cbarber) [inline app/javascript/components/widgets/weather/sunrise-sunset.jsx]: Specify the shape with `PropTypes.arrayOf(PropTypes.shape({ ...... }))`

File: app/javascript/components/widgets/weather/sunrise-sunset.jsx [hunk 15]
```diff
@@ -87,11 +89,11 @@ const SunriseSunset = ({ location }) => {
         timezone={timezone}
       />
       <SunriseLabel>
-        <Text>Sunrise</Text>
+        <Text>{isNight ? 'Sunset' : 'Sunrise'}</Text>
```
Comment (cbarber) [inline app/javascript/components/widgets/weather/sunrise-sunset.jsx]: Could we be cheeky and just:
```
const capitalize = R.replace(/^./, R.toUpper);
....
 <Text>{capitalize(startTime.type)}</Text>
```

or it could just be a lookup object.

File: app/javascript/components/widgets/weather/moon-phase.jsx (hunk 9 above)
Comment (cbarber) [reply app/javascript/components/widgets/weather/moon-phase.jsx]: if/else if/else conditions in that case. `switch` suggests an exact match for cases.

i suspect there is some refactoring opportunities here as well. Defining the propTypes for `phase` might help illustrate the options.

File: app/javascript/components/widgets/weather/semi-circle.jsx (hunk 11 above)
Comment (cbarber) [reply app/javascript/components/widgets/weather/semi-circle.jsx]: Yeah, stick with if/else if/else here as well.

File: app/models/location.rb [hunk 16]
```diff
@@ -34,5 +34,5 @@ def bathroom_code
     ENV['BATHROOM_CODE']
   end
 
-  has_many :solarcycles
+  has_many :solarcycles, dependent: :destroy
```
[the remark this answers is not in the document]
Comment (sracca) [reply app/models/location.rb]: Kind of, rubocop required a dependent field for solarcycles w/in location, so I thought destroy would be the best option in case of such an event.

File: app/controllers/web_hooks/solarcycles_controller.rb [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+  private def index
+    render
+  end
+
+  private def publish(event_scope)
+    event = event_scope.first_or_initialize
+    return unless event.new_record?
+
+    event.save!
+    Helios2Schema.subscriptions.trigger("eventPublished", {}, event)
```
[the remark this answers is not in the document]
Comment (sracca) [reply app/controllers/web_hooks/solarcycles_controller.rb]: whoops yea, that was me trying to figure out stuff and forgot to delete it. All set now!

File: app/assets/javascripts/solarcycles.coffee (hunk 1 above)
[the remark this answers is not in the document]
Comment (geraldalewis) [reply app/assets/javascripts/solarcycles.coffee]: 👹

File: app/javascript/components/widgets/weather/sunrise-sunset.jsx [hunk 18]
```diff
@@ -108,10 +122,16 @@ SunriseSunset.propTypes = {
       }),
     ),
   }).isRequired,
+  weather: PropTypes.shape({
+    daily: PropTypes.shape({
+      data: PropTypes.array.isRequired,
```
Comment (cbarber) [inline app/javascript/components/widgets/weather/sunrise-sunset.jsx]: Specify moonPhase proptype

**Comment by sracca** [comment]
I fixed all of the CircleCI issues plus the most recent suggestions, so everything should be all good!

File: app/javascript/components/widgets/weather/semi-circle.jsx [hunk 19]
```diff
@@ -7,6 +7,18 @@ import {
   timeAndDateForTimezone,
 } from '../../../lib/datetime';
 import { colors } from '../../../lib/theme';
+import MoonPhase from './moon-phase';
+
+const getPhaseType = phase => {
```
Comment (geraldalewis) [inline app/javascript/components/widgets/weather/semi-circle.jsx]: neat!
